In [32]:
%pip install folium pandas OSMnx networkx scikit-learn matplotlib numpy

Note: you may need to restart the kernel to use updated packages.


In [33]:
import solution_reader
import os
import pandas as pd
import numpy as np

In [ ]:
# Files

solution_directory = "./../output/selection_soSLNS"
best_known_file = "./../best_known/best_known_LL01_100_200.txt"

# Best known dataframe
best_known_df = solution_reader.read_best_known(best_known_file)

# Solution dataframe
solutions_list = []
files = os.listdir(solution_directory)
for solution_file in files :
    solution = solution_reader.PDPTWCompleteSolution(solution_directory + "/" + solution_file)
    
    solution_dict = {
    "Instance": solution.get_instance(),
    "Vehicles": solution.get_vehicles(),
    "Distance": solution.get_cost(),
    "Time": solution.get_time(),
    "Reference": solution.get_reference(),
    "Date": solution.get_date()
    }

    solutions_list.append(solution_dict)

solutions_df = pd.DataFrame(solutions_list)


best_known_df = best_known_df[best_known_df["Instance"].isin(solutions_df["Instance"])]
solutions_df.sort_values(by=["Instance"], inplace=True)
best_known_df.sort_values(by=["Instance"], inplace=True)

solutions_df.reset_index(drop=True, inplace=True)
best_known_df.reset_index(drop=True, inplace=True)

print(best_known_df)
print(solutions_df)

    Instance  Vehicles  Distance Reference       Date
0    LC1_2_1        20   2704.57    Li&Lim       2001
1    LC1_2_2        19   2764.56    Li&Lim       2001
2    LC1_2_3        17   3127.78         H  15-jul-16
3    LC2_2_1         6   1931.44    Li&Lim       2001
4    LC2_2_2         6   1881.40    Li&Lim       2001
5    LC2_2_3         6   1844.33  SAM::OPT  15-apr-03
6    LR1_2_1        20   4819.12    Li&Lim       2001
7    LR1_2_2        17   4621.21        RP  08-jul-03
8    LR1_2_3        14   4402.38       CLS  27-feb-16
9    LR2_2_1         5   4073.10  SAM::OPT  09-dec-03
10   LR2_2_2         4   3796.00  SAM::OPT  14-feb-03
11   LR2_2_3         4   3098.36        RP  08-jul-03
12  LRC1_2_1        19   3606.06  SAM::OPT  14-dec-03
13  LRC1_2_2        15   3671.02       EOE  19-nov-15
14  LRC1_2_3        13   3154.92       CLS  27-feb-16
15  LRC2_2_1         6   3595.18         K  14-apr-15
16  LRC2_2_2         5   3158.25         H  13-jul-16
17  LRC2_2_3         4   288

/home/a24jacqb/Documents/Code/pdptw-main/python/solution_reader.py:18: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  df = pd.read_csv(file, delim_whitespace=True, names=["Instance", "Vehicles", "Distance", "Reference", "Date"], skiprows=1)


In [35]:
sum_vehicles_best_known = np.sum(best_known_df["Vehicles"])
sum_vehicles_solution = np.sum(solutions_df["Vehicles"])
sum_time = np.sum(solutions_df["Time"])

sum_distance_best_known = np.sum(best_known_df["Distance"])
sum_distance_solution = np.sum(solutions_df["Distance"])

mean_vehicles_best_known = np.mean(best_known_df["Vehicles"])
mean_vehicles_solution = np.mean(solutions_df["Vehicles"])

mean_distance_best_known = np.mean(best_known_df["Distance"])
mean_distance_solution = np.mean(solutions_df["Distance"])

# Mean absolute error (MAE)
mae_vehicles = np.mean(np.abs(solutions_df["Vehicles"] - best_known_df["Vehicles"]))
mae_distances = np.mean(np.abs(solutions_df["Distance"] - best_known_df["Distance"]))

# Mean relative error (MRE)
mre_vehicles = np.mean(np.abs((solutions_df["Vehicles"] - best_known_df["Vehicles"]) / best_known_df["Vehicles"]))
mre_distances = np.mean(np.abs((solutions_df["Distance"] - best_known_df["Distance"]) / best_known_df["Distance"]))

# Mean ratio 
mr_vehicles = mean_vehicles_solution / mean_vehicles_best_known
mr_distances = mean_distance_solution / mean_distance_best_known

# Percentage error is calculated as (MAE / Actual value) * 100
percentage_error_vehicles = ((mean_vehicles_solution - mean_vehicles_best_known) / mean_vehicles_best_known) * 100
percentage_error_distances = ((mean_distance_solution - mean_distance_best_known) / mean_distance_best_known) * 100


print(f"Vehicle difference: {sum_vehicles_solution - sum_vehicles_best_known}")
print(f"Mean Absolute Error (distance): {mae_distances:.4f}")
print(f"Mean Relative Error (distance): {mre_distances:.4f}")
print(f"Mean Approximation Ratio (distance): {mr_distances:.4f}")
print(f"Percentage Error (vehicles): {percentage_error_vehicles:.2f}%")
print(f"Percentage Error (distances): {percentage_error_distances:.2f}%")
print(f"Total Time: {sum_time:.2f}")

Vehicle difference: 7
Mean Absolute Error (distance): 298.3672
Mean Relative Error (distance): 0.1376
Mean Approximation Ratio (distance): 1.0997
Percentage Error (vehicles): 2.03%
Percentage Error (distances): 9.97%
Total Time: 1898.00


Sort of SLNS
- Vehicle difference: 18
- Mean Absolute Error (distance): 103.3417
- Mean Relative Error (distance): 0.0327
- Mean Approximation Ratio (distance): 0.9587
- Percentage Error (vehicles): 5.23%
- Percentage Error (distances): -4.13%

SLNS
- Vehicle difference: 20
- Mean Absolute Error (distance): 105.9031
- Mean Relative Error (distance): 0.0334
- Mean Approximation Ratio (distance): 0.9588
- Percentage Error (vehicles): 5.81%
- Percentage Error (distances): -4.12%

Min + SLNS
- Vehicle difference: 7
- Mean Absolute Error (distance): 298.3672
- Mean Relative Error (distance): 0.1376
- Mean Approximation Ratio (distance): 1.0997
- Percentage Error (vehicles): 2.03%
- Percentage Error (distances): 9.97%



Old results (first selection)

5000 iterations
- Vehicle difference: 7
- Mean Absolute Error (distance): 26.8283
- Mean Relative Error (distance): 0.0231
- Mean Approximation Ratio (distance): 1.0033
- Percentage Error (vehicles): 4.86%
- Percentage Error (distances): 0.33%

10000 iterations
- Vehicle difference: 6
- Mean Absolute Error (distance): 21.3128
- Mean Relative Error (distance): 0.0188
- Mean Approximation Ratio (distance): 0.9985
- Percentage Error (vehicles): 4.17%
- Percentage Error (distances): -0.15%

negative % error due to some solution having more routes but less travel distances
